File: src/main/java/com/binarysushi/studio/webdav/StudioBulkFileListener.java
package com.binarysushi.studio.webdav;

import com.binarysushi.studio.settings.StudioSettingsProvider;
import com.intellij.openapi.Disposable;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.progress.PerformInBackgroundOption;
import com.intellij.openapi.progress.ProgressManager;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.project.ProjectManager;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.openapi.vfs.VirtualFileManager;
import com.intellij.openapi.vfs.newvfs.BulkFileListener;
import com.intellij.openapi.vfs.newvfs.events.VFileEvent;
import com.intellij.util.messages.MessageBusConnection;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public class StudioBulkFileListener implements BulkFileListener, Disposable {
    private MessageBusConnection connection;
    private static Logger LOG = Logger.getInstance(StudioBulkFileListener.class);

    public StudioBulkFileListener() {
        connection = ApplicationManager.getApplication().getMessageBus().connect();
        connection.subscribe(VirtualFileManager.VFS_CHANGES, this);
    }

    @Override
    public void dispose() {
        connection.disconnect();
    }

    @NotNull
    public String getComponentName() {
        return "StudioBulkFileListener";
    }

    @Override
    public void before(@NotNull List<? extends VFileEvent> events) {
    }

    @Override
    public void after(@NotNull List<? extends VFileEvent> events) {
        Project[] projects = ProjectManager.getInstance().getOpenProjects();

        for (VFileEvent event : events) {
            VirtualFile eventFile = event.getFile();

            if (eventFile != null && !eventFile.isDirectory()) {
                for (Project project : projects) {
                    StudioSettingsProvider settingsProvider = StudioSettingsProvider.getInstance(project);

                    // Bail out if auto uploads are not enabled.
                    if (!StudioSettingsProvider.getInstance(project).getAutoUploadEnabled()) {
                        return;
                    }

                    if (settingsProvider.getCartridgeRoots().size() < 1) {
                        return;
                    }

                    for (String cartridgeRoot : settingsProvider.getCartridgeRoots()) {
                        if (eventFile.getPath().contains(cartridgeRoot)) {
                            ProgressManager.getInstance().run(
                                    new StudioUpdateFileTask(
                                            project,
                                            "Syncing files to: " + StudioSettingsProvider.getInstance(project).getHostname(),
                                            true,
                                            PerformInBackgroundOption.ALWAYS_BACKGROUND,
                                            cartridgeRoot,
                                            eventFile.getPath()
                                    )
                            );
                        }
                    }
                }
            }
        }
    }
}

File: src/main/kotlin/com/binarysushi/studio/webdav/StudioBulkFileListener.kt
package com.binarysushi.studio.webdav

import com.binarysushi.studio.configuration.projectSettings.StudioConfigurationProvider
import com.intellij.openapi.progress.PerformInBackgroundOption
import com.intellij.openapi.progress.ProgressManager
import com.intellij.openapi.project.ProjectManager
import com.intellij.openapi.vfs.newvfs.BulkFileListener
import com.intellij.openapi.vfs.newvfs.events.VFileEvent

class StudioBulkFileListener : BulkFileListener {
    override fun after(events: List<VFileEvent>) {
        val projects = ProjectManager.getInstance().openProjects

        for (event in events) {
            if (event.file != null && !event.file!!.isDirectory) {
                for (project in projects) {
                    val configurationProvider = StudioConfigurationProvider.getInstance(project)

                    // Bail out if auto uploads are not enabled.
                    if (!StudioConfigurationProvider.getInstance(project).autoUploadEnabled) {
                        return
                    }

                    if (configurationProvider.cartridgeRoots.size < 1) {
                        return
                    }

                    for (cartridgeRoot in configurationProvider.cartridgeRoots) {
                        if (event.file!!.path.contains(cartridgeRoot)) {
                            val fileTask = StudioUpdateFileTask(
                                project,
                                "Syncing files to: " + StudioConfigurationProvider.getInstance(project).hostname,
                                true,
                                cartridgeRoot,
                                event.file!!
                            )

                            ProgressManager.getInstance().run(fileTask)
                        }
                    }
                }
            }
        }
    }
}
